Divide and Conquer Review

char-adadev opened the pull request:
The purpose of this PR is to pare down the Divide and Conquer lesson to be a review. The lesson was presented during the classroom phase previously, so most of the lesson is a recap of what they may have gone over during class. 

In addition to the review, there is an optional practice question added and two alternative challenges to check for understanding of the algorithms reviewed.

I'm happy to make the optional practice into a challenge question, lemme know your thoughts!

For review:
* Please run learn preview on the lesson
* Check for accuracy, correctness, grammar
* Please double check optional practice question solution for correctness
* Any and all suggestions welcome 🙏🏾

kyra-at-ada commented on 01-algorithms/02-algorithms-divide-conquer.md:
```diff
@@ -1,42 +1,24 @@
-When we understand the _kind_ of solution or algorithm we are using or can use, we can more easily read and understand it.
-
-There are a number of solutions and algorithms classified as using the divide-and-conquer approach. Learning about divide-and-conquer approaches will:
+You may recall learning about divide and conquer algorithms such as Binary Search, Quick Sort, and Merge Sort. This lesson is intended to serve as a refresher for the divide and conquer approach.
```
Move binary search to the end since it's _techically_ not divide and conquer.
```suggestion
You may recall learning about divide and conquer algorithms such as Quick Sort, and Merge Sort, Binary Search. This lesson is intended to serve as a refresher for the divide and conquer approach.
```

kyra-at-ada commented on 01-algorithms/02-algorithms-divide-conquer.md:
```diff
@@ -1,42 +1,24 @@
 
-<iframe src="https://adaacademy.hosted.panopto.com/Panopto/Pages/Embed.aspx?pid=04387545-7ced-4c84-9ce4-ad44002bde36&autoplay=false&offerviewer=true&showtitle=true&showbrand=false&start=0&interactivity=all" height="405" width="720" style="border: 1px solid #464646;" allowfullscreen allow="autoplay"></iframe>
-
 ## Goals
```
```suggestion
## Learning Goals
```

kyra-at-ada commented on 01-algorithms/02-algorithms-divide-conquer.md:
```diff
@@ -1,42 +1,24 @@
 ## Overview
 
-**Divide and Conquer** is an approach to problem solving that breaks down a large problem into multiple, smaller subproblems. We use the results of those subproblems to solve the original problem.
+As a reminder, **Divide and Conquer** is an approach to problem solving that breaks down a large problem into multiple, smaller subproblems. We use the results of those subproblems to solve the original problem.
```
```suggestion
As a reminder, **Divide and Conquer** is an approach to problem solving that breaks down a large problem into multiple, smaller subproblems. We combine the results of those subproblems to solve the original problem.
```

kyra-at-ada commented on 01-algorithms/02-algorithms-divide-conquer.md:
```diff
@@ -45,205 +27,15 @@ QuickSort is an algorithm which takes a divide-and-conquer approach to sorting a
 1. Move all elements smaller than the pivot to the left and all elements larger than the pivot to the right. Note that the pivot is now in the correct index. Picking a pivot and rearranging the elements is referred to as _partitioning_.
 1. Perform QuickSort on the left and right sides of the pivot.
```
I'm a visual learner so I feel like I'd benefit from adding the visualization back here

kyra-at-ada commented on 01-algorithms/02-algorithms-divide-conquer.md:
```diff
@@ -45,205 +27,15 @@ QuickSort is an algorithm which takes a divide-and-conquer approach to sorting a
 2. _Sort_ each subarray with the merge sort algorithm. (An array of size one is trivially sorted.)
 3. _Merge_ the subarrays into one array by combining two subarrays into one at each step.
 
-This process is usually done by keeping track of three indices in the array: _starting index_, _ending index_ and _midway index_ as shown in the image below.
```
Same here, I actually really like the photo we provide

char-adadev replied:
Added this back. Thank you for the suggestion!

kyra-at-ada commented on 01-algorithms/02-algorithms-divide-conquer.md:
```diff
@@ -257,35 +49,54 @@ Here's a description of the binary search algorithm:
+You are tasked with finding the smallest missing non-negative element inside of the array.
 
-    mid = (high + low) // 2
+The array should start with the integer 0, if it is present in the array.
```
```suggestion
If present, the array should start with the integer 0.
```

kyra-at-ada reviewed (commented):
Nice work! I like the added material. However, I just realized a major issue which is that we actually cut the divide and conquer/algorithmic strategies lesson from Accelerate because it's addressed in internship so those students haven't seen this problem before. 

Instead of making them read an extra lesson, I'm wondering if we could maybe walk through a new example of divide and conquer to illustrate more thoroughly how it works and provide fresher review content for everyone? I was thinking this problem: 
[Power of x to n](https://www.geeksforgeeks.org/write-a-c-program-to-calculate-powxn/) which again, can add the DP solution in the resources as a 'follow your curiosity'

or something even simpler like this: [find min/max element of array](https://www.techiedelight.com/find-minimum-maximum-element-array-minimum-comparisons/)

Thoughts?

char-adadev commented:
I updated the lesson with the min/max problem. Today I plan on creating visuals for the additional problems I added 🙏🏾 I'll re-request a review after to keep you updated.

kyra-at-ada commented on 01-algorithms/02-algorithms-divide-conquer.md:
```diff
@@ -46,246 +44,163 @@ QuickSort is an algorithm which takes a divide-and-conquer approach to sorting a
-_midway index_ = ⌊(_starting index_ + _ending index_) / 2⌋
-
-</div>
+Binary search can be considered a decrease-and-conquer algorithm because it decreases the size of the array by half with each step. No matter how large the original array was, at each step we'll be able to discard half the remaining data from consideration! The half-sized array will then be analyzed with the same step as the full-sized array.
```
```suggestion
Binary search is often called a decrease-and-conquer algorithm because it decreases the size of the array by half with each step. No matter how large the original array was, at each step we'll be able to discard half the remaining data from consideration! The half-sized array will then be analyzed with the same step as the full-sized array.
```

kyra-at-ada commented on 01-algorithms/02-algorithms-divide-conquer.md:
```diff
@@ -46,246 +44,163 @@ QuickSort is an algorithm which takes a divide-and-conquer approach to sorting a
+Binary search can be considered a decrease-and-conquer algorithm because it decreases the size of the array by half with each step. No matter how large the original array was, at each step we'll be able to discard half the remaining data from consideration! The half-sized array will then be analyzed with the same step as the full-sized array.
 
-Where ⌊⌋ denotes the mathematical _floor_ operation, or integer truncation.
+Binary search is technically not considered a divide-and-conquer algorithm but instead a decrease-and-conquer algorithm because divide-and-conquer algorithms typically have *two or more* subproblems which are generated from the main problem. In contrast, the binary search algorithm decreases to one sub-problem.
```
```suggestion
Binary search is technically not a divide-and-conquer algorithm but instead a decrease-and-conquer algorithm because divide-and-conquer algorithms typically have *two or more* subproblems which are generated from the main problem. In contrast, the binary search algorithm decreases to one subproblem.
```

kyra-at-ada commented on 01-algorithms/02-algorithms-divide-conquer.md:
```diff
@@ -46,246 +44,163 @@ QuickSort is an algorithm which takes a divide-and-conquer approach to sorting a
 
-For the first _divide_ step, the _midway index_ will be _(0+8)/2_ = _4_.
+1. Starting with a sorted array, we check whether it contains a particular value by comparing the search value with the element in the middle of the array.
+1. If we find the value we return the position where it was found.
```
Perhaps just me, but when I first read this step my brain wanted it to say 'we return _to_ the position', so removing the 'we' adds clarity. 
```suggestion
1. If we find the value, return the position where it was found.
```

kyra-at-ada replied:
```suggestion
1. If we find the value we return the position where the value was found.
```

kyra-at-ada commented on 01-algorithms/02-algorithms-divide-conquer.md:
```diff
@@ -46,246 +44,163 @@ QuickSort is an algorithm which takes a divide-and-conquer approach to sorting a
-For the first _divide_ step, the _midway index_ will be _(0+8)/2_ = _4_.
+1. Starting with a sorted array, we check whether it contains a particular value by comparing the search value with the element in the middle of the array.
+1. If we find the value we return the position where it was found.
+1. If we don't find the value, we determine whether it would be in the left or right half of the array by seeing whether it is smaller or larger than the middle element. Remember, we can make this decision because _we assume the array is sorted!_ Then we perform a binary search on the selected half.
```
```suggestion
1. If we don't find the value, we determine whether it would be in the left or right half of the array by checking whether it is smaller or larger than the middle element. Remember, we can make this decision because _we assume the array is sorted!_ Then we perform a binary search on the selected half.
```

kyra-at-ada replied:
```suggestion
1. If we don't find the value, we determine whether the value would be in the left or right half of the array by seeing whether the value is smaller or larger than the middle element. Remember, we can make this decision because _we assume the array is sorted!_ Then we perform a binary search on the selected half.
```

kyra-at-ada commented on 01-algorithms/02-algorithms-divide-conquer.md:
```diff
@@ -46,246 +44,163 @@ QuickSort is an algorithm which takes a divide-and-conquer approach to sorting a
+Here's a description of the binary search algorithm:
 
-For the first _divide_ step, the _midway index_ will be _(0+8)/2_ = _4_.
+1. Starting with a sorted array, we check whether it contains a particular value by comparing the search value with the element in the middle of the array.
```
Avoid ambiguous pronouns like 'it' when possible. 
```suggestion
1. Starting with a sorted array, we check whether the array contains a particular value by comparing the search value with the element in the middle of the array.
```

kyra-at-ada commented on 01-algorithms/02-algorithms-divide-conquer.md:
```diff
@@ -46,246 +44,163 @@ QuickSort is an algorithm which takes a divide-and-conquer approach to sorting a
+1. Starting with a sorted array, we check whether it contains a particular value by comparing the search value with the element in the middle of the array.
+1. If we find the value we return the position where it was found.
+1. If we don't find the value, we determine whether it would be in the left or right half of the array by seeing whether it is smaller or larger than the middle element. Remember, we can make this decision because _we assume the array is sorted!_ Then we perform a binary search on the selected half.
+1. If at any point, we end up with an empty range, we know the value was not in the array, and we can return a result indicating the value was not found, such as `None`. Other variations of binary search may return the index of where the value _should_ have been, but as a negative value to indicate that it was missing.
```
```suggestion
1. If at any point, we end up with an empty range, we know the value was not in the array, and we can return a result indicating the value was not found, such as `None`. Other variations of binary search may return the index of where the value _should_ have been, but as a negative value to indicate that the value was missing.
```

kyra-at-ada commented on 01-algorithms/02-algorithms-divide-conquer.md:
```diff
@@ -46,246 +44,163 @@ QuickSort is an algorithm which takes a divide-and-conquer approach to sorting a
+1. If we find the value we return the position where it was found.
+1. If we don't find the value, we determine whether it would be in the left or right half of the array by seeing whether it is smaller or larger than the middle element. Remember, we can make this decision because _we assume the array is sorted!_ Then we perform a binary search on the selected half.
+1. If at any point, we end up with an empty range, we know the value was not in the array, and we can return a result indicating the value was not found, such as `None`. Other variations of binary search may return the index of where the value _should_ have been, but as a negative value to indicate that it was missing.
+1. Each recursion divides the array in half and performs the binary search on a smaller subproblem.
```
```suggestion
1. Each recursive call divides the array in half and performs the binary search on a smaller subproblem.
```

kyra-at-ada commented on 01-algorithms/02-algorithms-divide-conquer.md:
```diff
@@ -46,246 +44,163 @@ QuickSort is an algorithm which takes a divide-and-conquer approach to sorting a
+The following material serves as an addendum to any Divide and Conquer content you may have covered in class. 
 
-### Sort and Merge
+## Minimum and Maximum Elements in an Array
```
Make subtopic of divide and conquer
```suggestion
### Minimum and Maximum Elements in an Array
```

kyra-at-ada commented on 01-algorithms/02-algorithms-divide-conquer.md:
```diff
@@ -46,246 +44,163 @@ QuickSort is an algorithm which takes a divide-and-conquer approach to sorting a
+1. Return the min and max values
 
-How do we efficiently merge two arrays? An auxiliary array of size _n_ is often used to facilitate the merge steps, after which the values are copied from the auxiliary array back to the original array.
+In this algorithm we are dividing the list up until there is only one element, making comparisons of the minimum and maximum values to this one element and updating the values if needed, and then combining the results by storing the minimum and maximum values encountered during our recursion.
```
```suggestion
In this algorithm we are dividing the list up until there is only one element, making comparisons of the minimum and maximum values to this one element and updating the values if needed, and then combining the results by storing the minimum and maximum values encountered during our recursive calls.
```

kyra-at-ada commented on 01-algorithms/02-algorithms-divide-conquer.md:
```diff
@@ -46,246 +44,163 @@ QuickSort is an algorithm which takes a divide-and-conquer approach to sorting a
+## Is there an echo in here?
 
-Since merge sort always divides the list in half at each divide step, there will be _log n_ levels in the division phase. By keeping data in place, and only adjusting indices to keep track of where the subarrays are located, at each level, there are at most _n_ calculations, for a time complexity of the division phase of _O(n log n)_.
+You may notice similarities between this algorithm and the Merge Sort algorithm described above. If so, that's great! The Merge Sort and Binary Search algorithms are excellent models for a large number of divide-and-conquer algorithms. Due to this fact, it is recommended to spend more time understanding Merge Sort and Binary Search than any other divide-and-conquer algorithms. 
```
Love this! Consider modifying to make it clear understanding the algorithms is the goal, not memorization. 
```suggestion
You may notice similarities between this algorithm and the Merge Sort algorithm described above. If so, that's great! The Merge Sort and Binary Search algorithms are excellent models for a large number of divide-and-conquer algorithms. While it is not necessary to memorize these algorithms, we do recommend prioritizing thoroughly understanding how the Merge Sort and Binary Search algorithms work over any other divide-and-conquer algorithms. 
```

kyra-at-ada commented on 01-algorithms/02-algorithms-divide-conquer.md:
```diff
@@ -46,246 +44,163 @@ QuickSort is an algorithm which takes a divide-and-conquer approach to sorting a
+### !end-callout
 
-Combining the divide and merge phases, we get a total complexity of _O(3n log n)_, which after dropping the coefficient gives us _O(n log n)_.
+Spend *no more than 15 minutes* writing a solution to the problem detailed above.
```
Consider adding a 0-points optional challenge question to give students a workspace

kyra-at-ada commented on 01-algorithms/02-algorithms-divide-conquer.md:
```diff
@@ -46,246 +44,163 @@ QuickSort is an algorithm which takes a divide-and-conquer approach to sorting a
+We can instead use a modified binary search algorithm to solve the problem in O(log(n)) time.
 
-Binary search can be considered a decrease-and-conquer algorithm because it decreases the size of the array by half with each step. No matter how large the original array was, at each step we'll be able to discard half the remaining data from consideration! The half-sized array will then be analyzed with the same step as the full-sized array.
+Spend *no more than 15 minutes* writing a modified binary search solution to the problem detailed above.
```
Again, I wonder if it would serve students to give them a workspace (via a challenge question) to do this work. Can be zero points so that it doesn't affect completion.

char-adadev replied:
Good idea, I'll add a challenge question for the two exercises

kyra-at-ada commented on 01-algorithms/02-algorithms-divide-conquer.md:
````diff
@@ -46,246 +44,163 @@ QuickSort is an algorithm which takes a divide-and-conquer approach to sorting a
+<summary>Click here to see a sample solution</summary>
 
-<summary>We can view an example of merge sort making use of index tracking by expanding this section.</summary>
+```py
````
💫 For whatever reason, I could not get code blocks within a details tab to register as code blocks so kudos to you for figuring that out

kyra-at-ada commented on 01-algorithms/02-algorithms-divide-conquer.md:
````diff
@@ -46,246 +44,163 @@ QuickSort is an algorithm which takes a divide-and-conquer approach to sorting a
-            merged.append(array[r])
-            r += 1
+```
+Suppose you are given a sorted array of non-negative distinct integers. 
````
Use we over you where possible. 
```suggestion
Suppose we are given a sorted array of non-negative distinct integers. 
```

kyra-at-ada commented on 01-algorithms/02-algorithms-divide-conquer.md:
````diff
@@ -46,246 +44,163 @@ QuickSort is an algorithm which takes a divide-and-conquer approach to sorting a
-            r += 1
+```
+Suppose you are given a sorted array of non-negative distinct integers. 
+You are tasked with finding the smallest missing non-negative element inside of the array.
````
```suggestion
We would like to find the smallest missing non-negative element inside of the array.
```

kyra-at-ada commented on 01-algorithms/02-algorithms-divide-conquer.md:
````diff
@@ -46,246 +44,163 @@ QuickSort is an algorithm which takes a divide-and-conquer approach to sorting a
 ```
 
-</details>
+This problem can be solved in O(n) time using a linear search to find the first index containing an element that does not match its value. However, this approach does not take advantage of the fact the input is sorted.
````
```suggestion
This problem can be solved in O(n) time using a linear search to find the first index containing an element whose value and index are not equal. However, this approach does not take advantage of the fact the input is sorted.
```

kyra-at-ada reviewed (commented):
Amazing updates Char! I think this strikes a really great balance between acknowledging many folks have seen this content before and giving enough detail to Accelerate folks who haven't encountered this before. 

I left a bunch of nitpicky comments style wise, and some suggestions about considering creating optional challenges in Learn so that students have a workspace to try out the last two problems (although I acknowledge that's a time intensive update). Really love the quick sort question asking them to think about time complexity.
